File: src/dartCode/bubble.dart
void main() {
  List<int> arr = [5, 3, -1, 1, 2, 1];
  insertion(arr);
  print(arr);
}

void insertion(List<int> arr){
for(int i = 0; i < arr.length -1; i++){
  for(int j = i+1; j> 0; j--){
    if(arr[j] < arr[j-1]){
      swap(arr, j, j-1);
    }
    else {
      break;
    }
  }
}
}

void selection(List<int> arr) {
  for (int i = 0; i < arr.length; i++) {
    // find the max item in the remaining array and swap with correct index
    int last = arr.length - i - 1;
    int maxIndex = getMaxIndex(arr, 0, last);
    swap(arr, maxIndex, last);
  }
}

void swap(List<int> arr, int first, int second) {
  int temp = arr[first];
  arr[first] = arr[second];
  arr[second] = temp;
}

int getMaxIndex(List<int> arr, int start, int end) {
  int max = start;
  for (int i = start; i <= end; i++) {
    if (arr[max] < arr[i]) {
      max = i;
    }
  }
  return max;
}

void bubble(List<int> arr) {
  late bool swapped;

  for (int i = 0; i < arr.length; i++) {
    swapped = false;
    for (int j = 1; j < arr.length - i; j++) {
      if (arr[j] < arr[j - 1]) {
        int temp = arr[j];
        arr[j] = arr[j - 1];
        arr[j - 1] = temp;
        swapped = true;
      }
    }
    if (!swapped) {
      break;
    }
  }
}

File: src/javaCode/Bubble.java
package javaCode;

import java.util.Arrays;

public class Bubble {
    public static void main(String[] args) {
        int[] arr = {5, 3, 4, 1, 2};
        selection(arr);
        System.out.println(Arrays.toString(arr));
    }

    static void selection(int[] arr){
        for (int i = 0; i < arr.length; i++) {
            // find the max item in the remaining array and swap with correct index
            int last = arr.length -i -1;
            int maxIndex = getMaxIndex(arr, 0, last);
            swap(arr, maxIndex, last);
        }
    }

    private static void swap(int[] arr, int first, int second) {
        int temp = arr[first];
        arr[first] = arr[second];
        arr[second] = temp;
    }


    private static int getMaxIndex(int[] arr, int start, int end) {
        int max =start;
        for (int i = start; i <= end ; i++) {
            if (arr[max] < arr[i]){
                max = i;
            }
        }
        return max;
    }

    static void bubble(int[] arr) {
        boolean swapped;
        //run the step n-1 times
        for (int i = 0; i < arr.length; i++) {
            swapped = false;
            // for each step, max item will come at last respective index
            for (int j = 1; j < arr.length - i; j++) {
                // swap if the item is smaller than the previous item
                if (arr[j] < arr[j-1]) {
                    //swap
                    int temp = arr[j];
                    arr[j] = arr[j-1];
                    arr[j-1] = temp;
                    swapped = true;
                }
            }
            // if you did not swap a particular value of i, it means the array is sorted hence stop the program
            if (!swapped) {
                break;
            }
        }
    }
}
